pkg/mimir/grpc_push_check.go

```diff
@@ -3,44 +3,124 @@
 package mimir
 
 import (
+	"context"
+	"net/http"
+	"strconv"
+
+	"github.com/grafana/dskit/grpcutil"
+	"github.com/grafana/dskit/httpgrpc"
 	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
 	"google.golang.org/grpc/status"
+
+	"github.com/grafana/mimir/pkg/api"
 )
 
-type pushReceiver interface {
+type ingesterPushReceiver interface {
 	StartPushRequest() error
 	FinishPushRequest()
 }
 
+// Interface exposed by Distributor.
+type distributorPushReceiver interface {
+	StartPushRequest(ctx context.Context, requestSize int64) (context.Context, error)
+	FinishPushRequest(ctx context.Context)
+}
+
 // getPushReceiver function must be constant -- return same value on each call.
-func newGrpcInflightMethodLimiter(getIngester func() pushReceiver) *grpcInflightMethodLimiter {
-	return &grpcInflightMethodLimiter{getIngester: getIngester}
+func newGrpcInflightMethodLimiter(getIngester func() ingesterPushReceiver, getDistributor func() distributorPushReceiver) *grpcInflightMethodLimiter {
+	return &grpcInflightMethodLimiter{getIngester: getIngester, getDistributor: getDistributor}
 }
 
 // grpcInflightMethodLimiter implements gRPC TapHandle and gRPC stats.Handler.
 type grpcInflightMethodLimiter struct {
-	getIngester func() pushReceiver
+	getIngester    func() ingesterPushReceiver
+	getDistributor func() distributorPushReceiver
 }
 
-const ingesterPushMethod = "/cortex.Ingester/Push"
+type ctxKey int
+
+const (
+	pushTypeCtxKey ctxKey = 1
+
+	ingesterPush    = 1
+	distributorPush = 2
+
+	ingesterPushMethod   string = "/cortex.Ingester/Push"
+	httpgrpcHandleMethod string = "/httpgrpc.HTTP/Handle"
+)
 
 var errNoIngester = status.Error(codes.Unavailable, "no ingester")
+var errNoDistributor = status.Error(codes.Unavailable, "no distributor")
 
-func (g *grpcInflightMethodLimiter) RPCCallStarting(methodName string) error {
-	if methodName == ingesterPushMethod {
+func (g *grpcInflightMethodLimiter) RPCCallStarting(ctx context.Context, methodName string, md metadata.MD) (context.Context, error) {
+	if g.getIngester != nil && methodName == ingesterPushMethod {
 		ing := g.getIngester()
 		if ing == nil {
 			// We return error here, to make sure that RPCCallFinished doesn't get called for this RPC call.
-			return errNoIngester
+			return ctx, errNoIngester
+		}
+
+		err := ing.StartPushRequest()
+		if err != nil {
+			return ctx, err
+		}
+
+		return context.WithValue(ctx, pushTypeCtxKey, ingesterPush), nil
+	}
+
+	if g.getDistributor != nil && methodName == httpgrpcHandleMethod {
+		// Let's check httpgrpc metadata, if present.
+		httpMethod := getSingleMetadata(md, httpgrpc.MetadataMethod)
+		httpUrl := getSingleMetadata(md, httpgrpc.MetadataURL)
+		msgSize := getMessageSize(md, grpcutil.MetadataMessageSize)
+
+		if httpMethod == http.MethodPost && (httpUrl == api.PrometheusPushEndpoint || httpUrl == api.OTLPPushEndpoint) {
```

Comment:
I'm not sure checking the URL for equality will work. Reason is that the endpoint is exposed under a configurable prefix (and I think the prefix is still the URL received here). What if we check if the URL ends with the requested endpoint? That's [what we do in the query-frontend](https://github.com/grafana/mimir/blob/dec27843609f58ecc0072e9cc222e126aee255e0/pkg/frontend/querymiddleware/roundtrip.go#L368-L382).

Reply:
Good point, done in https://github.com/grafana/mimir/pull/6300/commits/5cd002d94efdab78e581967f0d28b52ab344b61a.